scipy/_lib/_testutils.py

```diff
@@ -71,6 +73,73 @@ def __call__(self, label="fast", verbose=1, extra_argv=None, doctests=False,
         return (code == 0)
 
 
+class _TestPythranFunc:
+    '''
+    Inherit from this class to generate a bunch of relevant test cases
+    '''
+    ALL_INTEGER = [np.int8, np.int16, np.int32, np.int64]
+    ALL_FLOAT = [np.float32, np.float64]
+    ALL_COMPLEX = [np.complex64, np.complex128]
+
+    def setup_method(self):
+        self.dtypes = []
+        self.arguments = []
+        self.index = []
+        self.expected = None
+        self.func = None
+
+    def get_optional_args(self, func):
+        # get optional arguments with its default value,
+        # get the index of the first optional argument
+        signature = inspect.signature(func)
+        optional_args = {}
+        first_optional_arg_idx = None
+        for i, (k, v) in enumerate(signature.parameters.items()):
+            if v.default is not inspect.Parameter.empty:
+                optional_args[k] = v.default
+                if first_optional_arg_idx is None:
+                    first_optional_arg_idx = i
+        return optional_args, first_optional_arg_idx
+
+    def test_all_dtypes(self):
+        for dtype in self.dtypes:
+            # keep a copy of self.arguments to avoid in-place modify
+            test_array = self.arguments.copy()
+            for idx in self.index:
+                test_array[idx] = test_array[idx].astype(dtype)
+            self.pythranfunc(*test_array)
+
+    def test_views(self):
+        # keep a copy of self.arguments to avoid in-place modify
+        test_array = self.arguments.copy()
```

Comment:
These two lines are now present in every test case, however I don't see where the in-place modification would occur. Is this really necessary? No well-designed function should modify any of its input arguments.

Reply:
Ah.. I thought if I didn't do `test_array = self.arguments.copy()`, then `self. arguments` will be modified `self. arguments[idx] = self. arguments[idx][::-1][::-1]` ?  and I think the best way is to copy the array first, and then modified the some of the elements?

Reply:
Okay yes I see it now, sorry for the noise. One other suggestion, to avoid repeating the same comment everywhere: use a separate method, `args = self._get_args()`. With implementation:
```
def _get_args(self):
    # returns a copy of self.arguments because some of the test functions in this
    # class perform an in-place modification of the argument list before passing those
    # arguments to the function under test.
    return self.arguments.copy()